node/bft/src/helpers/telemetry.rs

```diff
@@ -0,0 +1,368 @@
+
+        // Insert the subdag certificates.
+        cfg_iter!(subdag).for_each(|(_round, certificates)| {
+            cfg_iter!(certificates).for_each(|certificate| {
+                // TODO (raychu86): Can be greatly optimized by doing a one-shot update instead of individual certificates.
+                self.insert_certificate(certificate);
+            })
+        });
+
+        // Calculate the participation scores.
+        self.update_participation_scores();
+    }
+
+    /// Insert a certificate to the telemetry tracker.
+    pub fn insert_certificate(&self, certificate: &BatchCertificate<N>) {
+        // Acquire the lock for `tracked_certificates`.
+        let mut tracked_certificates = self.tracked_certificates.write();
+
+        // Retrieve the certificate round, author, and ID.
+        let certificate_round = certificate.round();
+        let certificate_author = certificate.author();
+        let certificate_id = certificate.id();
+
+        // If the certificate already exists in the tracker, then return early.
+        if tracked_certificates.get(&certificate_round).map_or(false, |certs| certs.contains(&certificate_id)) {
+            return;
+        }
+
+        // Insert the certificate ID.
+        tracked_certificates.entry(certificate_round).or_default().insert(certificate_id);
+
+        // Acquire the lock for `validator_signatures`
+        let mut validator_signatures = self.validator_signatures.write();
+
+        // Insert the certificate author and signers.
+        for address in
+            [certificate_author].into_iter().chain(certificate.signatures().map(|signature| signature.to_address()))
+        {
+            validator_signatures
+                .entry(address)
+                .or_default()
+                .entry(certificate_round)
+                .and_modify(|count| *count += 1)
+                .or_insert(1);
+        }
+
+        // Acquire the lock for `validator_certificates`.
+        let mut validator_certificates = self.validator_certificates.write();
+
+        // Insert the certificate
+        validator_certificates.entry(certificate_author).or_default().insert(certificate_round);
+    }
+
+    /// Calculate and update the participation scores for each validator.
+    pub fn update_participation_scores(&self) {
+        // Fetch the certificates and signatures.
+        let tracked_certificates = self.tracked_certificates.read();
+        let validator_signatures = self.validator_signatures.read();
+        let validator_certificates = self.validator_certificates.read();
+
+        // Fetch the total number of certificates.
+        let num_certificate_rounds = tracked_certificates.len();
+        let total_certificates = validator_certificates.values().map(|rounds| rounds.len()).sum::<usize>();
+
+        // Calculate the signature participation scores for each validator.
+        let signature_participation_scores = Arc::new(RwLock::new(IndexMap::new()));
```

Comment:
also applies to the other 2 instances below: couldn't we use `collect` to populate these collections instead of locks and individual inserts? `indexmap` has a `rayon` feature, this should be possible

Reply:
Resolved with [674a213](https://github.com/ProvableHQ/snarkOS/pull/3516/commits/674a2139c08b1a64652310e17e870c46c05a6122)